File: app/src/main/java/android/mahendra/attendancemanager/viewmodels/SubjectListViewModel.java
package android.mahendra.attendancemanager.viewmodels;

import android.mahendra.attendancemanager.models.Subject;
import android.mahendra.attendancemanager.repositories.SubjectRepository;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import java.util.ArrayList;
import java.util.List;

public class SubjectListViewModel extends AndroidViewModel implements Subject.Callback {
    private static final String TAG = "SubjectListViewModel";

    private SubjectRepository mSubjectRepository;
    private LiveData<List<Subject>> mAllSubjects;

    public SubjectListViewModel(Application application) {
        super(application);
        mSubjectRepository = new SubjectRepository(application);
        mAllSubjects = mSubjectRepository.getAllSubjects();
    }

    public LiveData<List<Subject>> getAllSubjects() {
        return mAllSubjects;
    }

    public void insert(Subject subject) {
        mSubjectRepository.insert(subject);
    }

    public void update(Subject subject) {
        mSubjectRepository.update(subject);
    }

    @Override
    public void onAttended(Subject subject) {
        update(subject);
    }

    @Override
    public void onMissed(Subject subject) {
        update(subject);
    }

    @Override
    public void onCancelled(Subject subject) {
        update(subject);
    }

    private List<String> extractTitles(List<Subject> subjects) {
        List<String> subjectTitles = new ArrayList<>();
        for (Subject subject : subjects) {
            subjectTitles.add(subject.getTitle());
        }
        return subjectTitles;
    }

    public LiveData<List<String>> getSubjectTitles() {
        return Transformations.map(mAllSubjects, this::extractTitles);
    }
}

File: app/src/main/java/android/mahendra/attendancemanager/viewmodels/SubjectListViewModel.kt
package android.mahendra.attendancemanager.viewmodels

import android.mahendra.attendancemanager.models.Subject
import android.mahendra.attendancemanager.repositories.SubjectRepository
import androidx.lifecycle.*

import kotlinx.coroutines.launch

import kotlin.collections.List
import kotlin.collections.ArrayList

class SubjectListViewModel internal constructor(
    private val subjectRepository: SubjectRepository,
    private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    val allSubjects: LiveData<List<Subject>> = subjectRepository.allSubjects

    fun insert(subject: Subject) = viewModelScope.launch {
        subjectRepository.insert(subject)
    }

    fun update(subject: Subject) = viewModelScope.launch {
        subjectRepository.update(subject)
    }

    fun delete(subject: Subject) = viewModelScope.launch {
        subjectRepository.delete(subject)
    }

    fun onResetAttendance(subject: Subject) = viewModelScope.launch {
        subject.missedClasses = 0
        subject.attendedClasses = 0
        subjectRepository.update(subject)
    }

    private fun extractTitles(subjects: List<Subject>): List<String> {
        val subjectTitles = ArrayList<String>()
        for (subject in subjects) {
            subjectTitles.add(subject.title)
        }
        return subjectTitles
    }

    fun onUpdateTitle(oldTitle: String, newTitle: String) = viewModelScope.launch {
        subjectRepository.updateTitle(oldTitle, newTitle)
    }

    fun getSubjectTitles(): LiveData<List<String>> {
        return Transformations.map(allSubjects, this::extractTitles)
    }

    fun onDeleteSubjectWith(subjectTitle: String) = viewModelScope.launch {
        val subject = subjectRepository.getSubject(subjectTitle)
        subjectRepository.delete(subject)
    }

    fun onResetAttendance(subjectTitle: String) = viewModelScope.launch {
        val subject = subjectRepository.getSubject(subjectTitle)
        subject?.attendedClasses = 0
        subject?.missedClasses = 0
        subjectRepository.update(subject)
    }

    fun getSubject(subjectTitle: String): Subject? {
        return subjectRepository.getSubject(subjectTitle)
    }

    fun onNewSubject(subjectTitle: String) = viewModelScope.launch {
        subjectRepository.insert(Subject(subjectTitle))
    }

    fun onUpdateAttendance(subjectTitle: String, attendedClasses: Int, totalClasses: Int) = viewModelScope.launch {
        val missedClasses = totalClasses - attendedClasses
        subjectRepository.updateAttendance(subjectTitle, attendedClasses, missedClasses)
    }
}
